File: src/main/java/org/jetbrains/pmdkotlin/lang/kotlin/KotlinTokenManager.java
package org.jetbrains.pmdkotlin.lang.kotlin;

import com.intellij.openapi.util.TextRange;
import com.intellij.psi.PsiElement;
import com.intellij.psi.tree.IElementType;
import com.intellij.psi.util.PsiTreeUtil;
import net.sourceforge.pmd.lang.TokenManager;
import org.jetbrains.kotlin.lexer.KotlinLexer;
import org.jetbrains.kotlin.lexer.KtTokens;
import org.jetbrains.kotlin.psi.KtTypeElement;
import org.jetbrains.pmdkotlin.cpd.KotlinToken;

public class KotlinTokenManager extends KotlinLexer implements TokenManager, KtTokens {
    private static ThreadLocal<String> fileName = new ThreadLocal();
    public KotlinFileContext kotlinFileContext;

    public KotlinTokenManager() {
        super();
    }

//    public KotlinTokenManager(Reader source) {
//        super();
//        try {
//            String s = IOUtils.toString(source);
//            start(s, 0, s.length(), 0);
//        } catch (IOException e) {
//            e.printStackTrace();
//        }
//    }

    public KotlinTokenManager(KotlinFileContext kotlinFileContext) {
        this.kotlinFileContext = kotlinFileContext;
        String src = this.kotlinFileContext.sourceCodeToString();
        start(src, 0, src.length(), 0);
    }

    public Object getCurrentToken() {
        IElementType tokenType = getTokenType();
        if (tokenType != null) {
            String image = getBufferSequence().subSequence(getTokenStart(), getTokenEnd()).toString();
            TextRange range = new TextRange(getTokenStart(), getTokenEnd());

            return new KotlinToken(image, kotlinFileContext.filename, getTokenStart(), kotlinFileContext.getBeginLine(range), tokenType);
        }
        return null;
    }

    public Object getNextToken() {
        advance();
        return getCurrentToken();
    }

    public void setFileName(String fileName_) {
        fileName.set(fileName_);
    }

    public static String getFileName() {
        String fileName_ = fileName.get();
        return fileName_ == null ? "(no file name provided)" : fileName_;
    }

    public PsiElement findElementAt(int offset) {
        return kotlinFileContext.psiFile.findElementAt(offset);
    }

    public boolean isTypeToken(KotlinToken token) {
        PsiElement leaf = findElementAt(token.ofset);
        return PsiTreeUtil.getParentOfType(leaf, KtTypeElement.class) != null;
    }
}

File: src/main/kotlin/org/jetbrains/pmdkotlin/lang/kotlin/KotlinTokenManager.kt
package org.jetbrains.pmdkotlin.lang.kotlin

import com.intellij.openapi.util.TextRange
import com.intellij.psi.PsiElement
import com.intellij.psi.util.PsiTreeUtil
import net.sourceforge.pmd.lang.TokenManager
import org.jetbrains.kotlin.lexer.KotlinLexer
import org.jetbrains.kotlin.psi.KtTypeElement
import org.jetbrains.pmdkotlin.cpd.KotlinToken

class KotlinTokenManager(val kotlinFileContext: KotlinFileContext) : KotlinLexer(), TokenManager {

    var theFileName = "n/a"

    init {
        val src = reduceWindowsLineEndings(this.kotlinFileContext.sourceCodeToString())
        start(src, 0, src.length, 0)
    }

    val currentToken: KotlinToken?
        get() {
            val tokenType = tokenType
            if (tokenType != null) {
                val image = bufferSequence.subSequence(tokenStart, tokenEnd).toString()
                val range = TextRange(tokenStart, tokenEnd)

                return KotlinToken(image, kotlinFileContext.filename, tokenStart, kotlinFileContext.getBeginLine(range), tokenType)
            }
            return null
        }

    override fun getNextToken(): KotlinToken? {
        advance()
        return currentToken
    }

    override fun setFileName(fileName: String) {
        theFileName = fileName
    }

    fun findElementAt(offset: Int): PsiElement? {
        return kotlinFileContext.psiFile.findElementAt(offset)
    }

    fun isTypeToken(token: KotlinToken): Boolean {
        val leaf = findElementAt(token.ofset)
        return PsiTreeUtil.getParentOfType(leaf, KtTypeElement::class.java) != null
    }

    /**
     * Workaround for double-counted line-endings on windows environments.
     * https://youtrack.jetbrains.com/issue/KT-13489
     */
    private fun reduceWindowsLineEndings(code: String): String = code.replace("\r\n", "\n")
}
